File: tsfile/src/main/java/org/apache/iotdb/tsfile/file/metadata/ChunkMetaData.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.iotdb.tsfile.file.metadata;

import org.apache.iotdb.tsfile.file.metadata.enums.TSDataType;
import org.apache.iotdb.tsfile.file.metadata.statistics.Statistics;
import org.apache.iotdb.tsfile.utils.ReadWriteIOUtils;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.ByteBuffer;
import java.util.Objects;

/**
 * MetaData of one chunk.
 */
public class ChunkMetaData {

  private static final Logger LOG = LoggerFactory.getLogger(ChunkMetaData.class);


  private String measurementUid;

  /**
   * Byte offset of the corresponding data in the file Notice: include the chunk header and marker.
   */
  private long offsetOfChunkHeader;

  private TSDataType tsDataType;

  /**
   * version is used to define the order of operations(insertion, deletion, update). version is set
   * according to its belonging ChunkGroup only when being queried, so it is not persisted.
   */
  private long version;

  /**
   * All data with timestamp <= deletedAt are considered deleted.
   */
  private long deletedAt = Long.MIN_VALUE;

  private Statistics statistics;

  private ChunkMetaData() {
  }

  /**
   * constructor of ChunkMetaData.
   *
   * @param measurementUid measurement id
   * @param tsDataType time series data type
   * @param fileOffset file offset
   * @param statistics value statistics
   */
  public ChunkMetaData(String measurementUid, TSDataType tsDataType, long fileOffset,
    Statistics statistics) {
    this.measurementUid = measurementUid;
    this.tsDataType = tsDataType;
    this.offsetOfChunkHeader = fileOffset;
    this.statistics = statistics;
  }

  @Override
  public String toString() {
    return String.format("numPoints %d", statistics.getCount());
  }

  public long getNumOfPoints() {
    return statistics.getCount();
  }

  /**
   * get offset of chunk header.
   *
   * @return Byte offset of header of this chunk (includes the marker)
   */
  public long getOffsetOfChunkHeader() {
    return offsetOfChunkHeader;
  }

  public String getMeasurementUid() {
    return measurementUid;
  }

  public Statistics getStatistics() {
    return statistics;
  }

  public long getStartTime() {
    return statistics.getStartTime();
  }

  public long getEndTime() {
    return statistics.getEndTime();
  }

  public TSDataType getDataType() {
    return tsDataType;
  }

  /**
   * serialize to outputStream.
   *
   * @param outputStream outputStream
   * @return length
   * @throws IOException IOException
   */
  public int serializeTo(OutputStream outputStream) throws IOException {
    int byteLen = 0;

    byteLen += ReadWriteIOUtils.write(measurementUid, outputStream);
    byteLen += ReadWriteIOUtils.write(offsetOfChunkHeader, outputStream);
    byteLen += ReadWriteIOUtils.write(tsDataType, outputStream);
    byteLen += statistics.serialize(outputStream);
    return byteLen;
  }

  /**
   * deserialize from ByteBuffer.
   *
   * @param buffer ByteBuffer
   * @return ChunkMetaData object
   */
  public static ChunkMetaData deserializeFrom(ByteBuffer buffer) {
    ChunkMetaData chunkMetaData = new ChunkMetaData();

    chunkMetaData.measurementUid = ReadWriteIOUtils.readString(buffer);
    chunkMetaData.offsetOfChunkHeader = ReadWriteIOUtils.readLong(buffer);
    chunkMetaData.tsDataType = ReadWriteIOUtils.readDataType(buffer);

    chunkMetaData.statistics = Statistics.deserialize(buffer, chunkMetaData.tsDataType);

    return chunkMetaData;
  }

  public long getVersion() {
    return version;
  }

  public void setVersion(long version) {
    this.version = version;
  }

  public long getDeletedAt() {
    return deletedAt;
  }

  public void setDeletedAt(long deletedAt) {
    this.deletedAt = deletedAt;
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    ChunkMetaData that = (ChunkMetaData) o;
    return offsetOfChunkHeader == that.offsetOfChunkHeader &&
        version == that.version &&
        deletedAt == that.deletedAt &&
        Objects.equals(measurementUid, that.measurementUid) &&
        tsDataType == that.tsDataType &&
        Objects.equals(statistics, that.statistics);
  }
}

File: tsfile-go/src/tsfile/file/metadata/ChunkMetaData.go
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package metadata

import (
	"bytes"
	_ "log"
	"tsfile/common/constant"
	"tsfile/common/utils"
)

type ChunkMetaData struct {
	sensor                        string
	fileOffsetOfCorrespondingData int64
	numOfPoints                   int64
	totalByteSizeOfPagesOnDisk    int64
	startTime                     int64
	endTime                       int64
	valuesStatistics              *TsDigest
}

func (c *ChunkMetaData) Sensor() string {
	return c.sensor
}

func (c *ChunkMetaData) TotalByteSizeOfPagesOnDisk() int64 {
	return c.totalByteSizeOfPagesOnDisk
}

func (c *ChunkMetaData) FileOffsetOfCorrespondingData() int64 {
	return c.fileOffsetOfCorrespondingData
}

func (f *ChunkMetaData) Deserialize(reader *utils.BytesReader) {
	f.sensor = reader.ReadString()
	f.fileOffsetOfCorrespondingData = reader.ReadLong()
	f.numOfPoints = reader.ReadLong()
	f.totalByteSizeOfPagesOnDisk = reader.ReadLong()
	f.startTime = reader.ReadLong()
	f.endTime = reader.ReadLong()

	digest := new(TsDigest)
	digest.Deserialize(reader)

	f.valuesStatistics = digest
}

func (f *ChunkMetaData) GetSerializedSize() int {
	size_statistics := 4
	if f.valuesStatistics != nil {
		size_statistics = f.valuesStatistics.GetSerializedSize()
	}

	return constant.INT_LEN + len(f.sensor) + 5*constant.LONG_LEN + size_statistics
}

func (t *ChunkMetaData) SetDigest(tsDigest *TsDigest) {
	t.valuesStatistics = tsDigest
}

func (t *ChunkMetaData) GetStartTime() int64 {
	return t.startTime
}

func (t *ChunkMetaData) GetEndTime() int64 {
	return t.endTime
}

func (t *ChunkMetaData) SetTotalByteSizeOfPagesOnDisk(size int64) {
	t.totalByteSizeOfPagesOnDisk = size
}

func (t *ChunkMetaData) SetNumOfPoints(num int64) {
	t.numOfPoints = num
}

func (t *ChunkMetaData) SerializeTo(buf *bytes.Buffer) int {
	var byteLen int

	n1, _ := buf.Write(utils.Int32ToByte(int32(len(t.sensor)), 0))
	byteLen += n1
	n2, _ := buf.Write([]byte(t.sensor))
	byteLen += n2

	n3, _ := buf.Write(utils.Int64ToByte(t.fileOffsetOfCorrespondingData, 0))
	byteLen += n3
	n4, _ := buf.Write(utils.Int64ToByte(t.numOfPoints, 0))
	byteLen += n4
	n5, _ := buf.Write(utils.Int64ToByte(t.totalByteSizeOfPagesOnDisk, 0))
	byteLen += n5
	n6, _ := buf.Write(utils.Int64ToByte(t.startTime, 0))
	byteLen += n6
	n7, _ := buf.Write(utils.Int64ToByte(t.endTime, 0))
	byteLen += n7

	if t.valuesStatistics.sizeOfList <= 0 {
		byteLen += t.valuesStatistics.GetNullDigestSize()
	} else {
		// tsDigest serializeTo
		byteLen += t.valuesStatistics.serializeTo(buf)
	}

	return byteLen
}

func NewTimeSeriesChunkMetaData(sid string, fOffset int64, sTime int64, eTime int64) (*ChunkMetaData, error) {
	return &ChunkMetaData{
		sensor: sid,
		fileOffsetOfCorrespondingData: fOffset,
		startTime:                     sTime,
		endTime:                       eTime,
		totalByteSizeOfPagesOnDisk:    0,
		numOfPoints:                   0,
	}, nil
}
